### ISSUE TITLE
Improve performance on localization file parsing

### ISSUE THREAD

**Issue by Mte90** [WordPress/performance#171]
## Feature Description

Ref: https://core.trac.wordpress.org/ticket/17268

The issue now is that for every file mo is read by PHP with a binary parser for every single page request and there isn't any object cache or using the native PHP module that exists since php 5.3 (I am not sure but it was already avalaible in 2017 when I did my patch).

More then half of the wordpress websites are not in english -> https://wordpress.org/about/stats/ it is something that can have an effect in a lot of things including rest endpoints (for gutenberg in 5.0 it was implemented a code that doesn't load a mo file if is not needed in the page to improve the performance).

5 years ago (before php 7) I did some comparison with a patch on wp 4.9 -> https://imgur.com/a/4k9oi (you find everything in the ticket) 500 vs 800 ms with some booster with object cache

**Comment by eclarke1** [comment]
@tillkruss is this new issue something that should sit inside the Object Cache focus area? If so, I can add it to there for us and add to the Backlog for your prioritisation.

**Comment by lkraav** [comment]
Linking in case everybody here doesn't know about https://github.com/inpsyde/translation-cache yet.

**Comment by PhilMakower** [comment]
Also
https://github.com/polylang/dynamo
Which is the recommended "successor" to mu plugin dynamic-no-loader
https://github.com/aucor/dynamic-mo-loader

**Comment by lynt-smitka** [comment]
I made 2 simple MU plugins many years ago to solve this issue. Hundreds sites use it in production without issues.

The first one isn't dependent on anything else, it saves the processed localization object into PHP file - it prevents heavy serialization and it can be cached via PHP OPcache.

https://github.com/lynt-smitka/WP-nginx-config/blob/master/extras/mu-plugins/lynt-mo-cache.php

The second one use igbinary PHP extension (I'm huge fan of igbinary) - very effective binary serializer. It is even a little bit faster.

https://github.com/lynt-smitka/WP-nginx-config/blob/master/extras/mu-plugins/lynt-mo-cache-igb.php


Some benchmarks:
https://www.slideshare.net/vsmitka/wordcamp-bratislava-2019-cache/31

1 - no caching at all
2 - OPcache enabled
3 - Localization cache via PHP file objects
4 - Localization cache via igbinary
5 - Usage of PHP file objects as Object Cache

**Comment by Bjoern76** [comment]
8 years ago I extended the original native gettext implementation from the referenced ticket, made it a plugin (https://wordpress.org/plugins/wp-performance-pack/) which includes my original dynamic mo reader - which I meanwhile found used by others without any credit, which is kind of sad, but not as sad as the fact that the original ticket is still open even though me and many others like @Mte90 offered working solutions. I gave up and am just using my plugin which is still working fine on all sites I  maintain.

But to stay on topic: This issue definitely doesn't belong into any caching focus. The basic issue is that WP in its current implementation is **always** loading **all** translations on every page load. This eats up memory and CPU. Just moving to caching doesn't change much except from where the translations are loaded.

To improve on this you could cache the translations needed for each different page call. I implemented this in my plugin if you enabled the use of object caching. But as you have to create different cache entries for each page this can eat up cache space quite fast. I tried to optimize cache usage by using only one cache entry for the whole front end and a base cache for all admin pages and additional cache entries for different admin pages if needed.

But caches must get build and updated. And each time this has to be done the old inefficient translation mechanism will be used. Thats why I wrote the dynamic mo reader. mo files are created to be uses dynamically. So instead of parsing the whole file with each page load only when a translation is requested it is "partially parsed on demand". Even though broken down to a single translation lookup this is slower, it is vastly faster in whole, as each page load only uses a few translationbs from only some mo files.

So now not every registered mo file is loaded and parsed on each page load, but only those really needed and from those files only the requested translations which brings us to the last WP issue: WP is translating way too much strings on each load that aren't actually used. Each transaltion is still a costly call and WP translates many strings each call like month and day names, a bunch of script strings and many other strings. I tried to adress this (not so easy when it has to work as plugin) in my "JIT script translation" by implementing objects which behave like arrays using getters and setters and translating the strings not when they are added but when the getter is called. This further improves performance as unnecessary translation calls are reduced (but requires regular updates when implemented as a plugin).

_TL/DR: WP loads too many translations too often in an inefficient way. For better performance: Reduce translation calls, load only what is used, translate (i.e. parse) only what is used, caching should be optional._

Feel free to use any part of my plugin code with or without credit, I don't care.

**Comment by Mte90** [comment]
I am here to ping again 😁

I am available to join a performance chat meeting if it is on schedule to discuss about this as the [trac ticket](https://core.trac.wordpress.org/ticket/17268#comment:105) is marked for WP 6.1 and I am very tired of this issue and all the time required just to discuss about... nothing at the end.

As I said at various WordCamp (when they were live) I will eat a pineapple pizza 😅 if this ticket is addressed and resolved if can help to finally include something for this (see an italian eating a pineapple pizza is something that for our grandmothers is like heartbreaking).

WordPress cares about diversity but not about if the technology offers the same quality in the diversity field, as we are talking about a **discrimination** for instances that don't use the default locale `en_US`.
I want to remember that means the 56% of the WP websites globally https://wordpress.org/about/stats/ not just some angry developers like me 🤪

**Comment by dd32** [comment]
> @Bjoern76 said:
> So now not every registered mo file is loaded and parsed on each page load, but only those really needed and from those files only the requested translations which brings us to the last WP issue: WP is translating way too much strings on each load that aren't actually used.

IMHO; this is the major point that's worth making.
WordPress has added lazy-loading/just-in-time-loading, etc, but all that gains very little when the files ultimately have to be loaded anyway due to translations being required at some point of the pageload.
Optimisations in the `.mo` parsing and loading only gains so much, and changes in the core PHP runtime around object/array/strings has made a larger impact between PHP5->7->8 than we could have ever done by optimizing the reader code.

WordPress is hamstrung at a technical level by the fact that strings are usually translated far before it's known if it'll be displayed.
For example;
 - The `labels` of [register_post_type()](https://developer.wordpress.org/reference/functions/register_post_type/) will likely never be displayed on the front-end, but due to the `__()` translation calls occurring during the init of the pageload  the translation files are loaded and translated, but the resulting string will likely never be used.
 - The human-readable schema fields for the rest api will be provided in the initial [register_rest_route()](https://developer.wordpress.org/reference/functions/register_rest_route/) call, but will never be seen by a end-user in the vast majority of rest-api requests, as it'll be returning the requested data, not the schema.

This is always going to be the case, and attempts to avoid that are just going to be pointless without major changes how how WordPress works, that is not reasonably a viable option.

---

Using object-cache caching of POMO data has also tried by some (over and over again), and has showed benefits in some environments but experience of others pointed out that storing (in some cases) megabytes of data in a remote key-value store can result in poorer performance than file parsing, which is why the notion of caching has been dismissed by many existing core contributors over the years.

[7 years ago](https://github.com/dd32/ginger-mo), I started working on an alternative implementation of a `.mo` reader, the performance was reasonably good compared to what core had at the time, but some optimizations were made to the existing pomo classes which gained some of the benefits, and sped things up enough that I got no traction in the alternative implementation. One of the core things I had focused on was supporting existing `.mo` files (because WordPress can't drop support for it), the new-at-the-time `.json` jed translations (which many thought would be more efficient and safer than a binary format), and [my preferred `.php` format](https://github.com/dd32/ginger-mo/blob/master/tests/data/example-simple.php) which just skipped the file parsing step, accepting the risk of malicious php in the file in return for the speed/opcode benefits.


So... when you take into account the limitations WordPress is working within, how PHP works, how things are written in WordPress (not using a MVP style for example), and how translations are stored..
_IMHO_ the biggest benefit that we can realistically see for translation performance, is to drop the `.mo` format in favour of `.php` files; doing so would remove the overhead of parsing pomo files, remove the need to use PHP to parse files, rely upon PHP's builtin parser, gain PHP's opcode caching and filestat caching, and not be require attempting to change all of the WordPress functions which accept translated strings as args. It has some con's, notably exhaustion of opcache, but that's not out problem as such.

I should also add - IMHO using `.php` files shouldn't be done as a "caching layer" but rather as a file format delivered by WordPress.org alongside the `.mo, .po` files and ultimately replacing them.

WordPress hasn't intentionally caused localised sites to run slower,  it's just the unfortunate outcome of how the WordPress internationalisation community has grown over the last decade, and how hard it is to change something so complex (yet simple) and core-to WordPress when there isn't wide-spread agreement or people who are willing to work on it _at any specific point in time_ (or worse, both happen and it ends up with [too many cooks in the kitchen](https://knowyourphrase.com/too-many-cooks-in-the-kitchen) nothing moves forward..).

(Having this ticket open here, with discussions, in addition to [#17268](https://core.trac.wordpress.org/ticket/17268) really just means this discussion is even more complicated, as now anyone has to read the same stuff in multiple places)

**Comment by Mte90** [comment]
I know about the issue of too many things to read, I did a recap in the ticket a month ago and here for that reason.

Talking in the PHP-FIG group about the new standard they are evaluating for translation files seems that the option are two as you said: php or json files.

Maybe the best solution is to provide json files that will generate in wordpress the php files so there aren't less issues on providing malicious php files. A unique php file per language with all the strings in the website so can be cached more easily and updated on new files available or plugin updates.

At same time I think that something for mo files to speed up their usage is important as it will take a lot for their deprecation around.

I don't know what are honestly the next steps to moving further the idea. 

- Do a feature plugin in a new repository? 
- Create a survey to decide what to do?
- Create a shared document with a full recap?

Otherwise we will never finish the pre discussion part for the new implementation.

PS: https://projectfluent.org/ is by mozilla to create a new language format but I don't think there is a [php library](https://github.com/tacoberu/fluent-php) complete and is useful to us (it is a complete new system and is good as fallback).

**Comment by dd32** [comment]
> Maybe the best solution is to provide json files that will generate in wordpress the php files

While at first it seems obvious to just let WordPress generate the .php files, there are cases where WordPress is deployed in such a way that the production system has no write-access to files, where any automated "cache into .php files" code just won't work. 

> so there aren't less issues on providing malicious php files

Personally, I don't consider that to be an issue, as WordPress.org already provides plugin/theme packages that are often auto-updated, and in the [past code was executed directly from the `.mo` files anyway as PHP](https://core.trac.wordpress.org/ticket/41562). Yeah, it would be nice to avoid `.php` but if the performance benefit is worth it..

> At same time I think that something for mo files to speed up their usage is important as it will take a lot for their deprecation around.

I disagree, at least for the initial work, if part of the solution is to introduce a different storage format.
WordPress.org provides the majority of translations used by WordPress, so the task can be simplified by leaving `.mo` usage and performance alone at first and focusing on simply making things better going forward, and as sites upgrade and newer translation packages are built, things will get better.
Once that's complete, then further optimisations to the existing `.mo` formats can be considered.

That's assuming that the changes required for a performance improvement in `.mo` isn't a _small change_ and requires _a lot_ of alterations that is hard to review.

> Talking in the PHP-FIG group about the new standard they are evaluating for translation files seems that the option are two as you said: php or json files.

What about a third option, to throw a curve ball in, and say `.ini` files? It's plain-text, non-executable, PHP has a C-based parser included in PHP4+ `parse_ini_file()`, supports inline comments, is human readable, supports sections so metadata and strings can be split, is already used for translations in other projects (although, as `COLOR = "Colour"` not as `Color = Colour`), and could be generated along side `.po, .mo` files by WordPress.org packagers..

Aka, pretty much the same as `.json` files, except it has a simpler parser, supports inline comments, and a _little_ more human-editable.. 

Main downside is that it would be yet-another format.. although one that as said, is kind-of already used in other systems..

> I don't know what are honestly the next steps to moving further the idea.
> * Do a feature plugin in a new repository?

I would personally recommend a feature plugin, but, that also requires multiple people to agree upon a way forward. And as said, there's multiple people with multiple approaches suggested with different opinions and ideals, how I would approach this is likely different to you, is different than what SomeOtherPerson would.

As you said, pre-discussion will ultimately eat everyones time, so the real solution is for someone to objectively weigh up all the existing discussions, and approaches, and find a middle-ground that is doable and is scoped for the main goal than changing everything at once.

WordPress is an opinionated piece of software though, so the solutions just need to be something that aligns with the opinions and experience of the [Internationalization component maintainers](https://make.wordpress.org/core/components/i18n/), who are the most likely people who would review/commit any work here.

> PS: https://projectfluent.org/ is by mozilla to create a new language format

Fluent looks interesting, but, it doesn't appear that it would be a very easy drop-in replacement for existing gettext based translations, or at least would result in skipping some of it's benefits for compatibility, and the flexibility would likely result in worse translation performance (once again, like said, due to WordPress translating everything even when not displayed)

**Comment by Mte90** [comment]
Just to keep the discussion open https://github.com/wearerequired/translations-cache in this case a cache solution for language files.

I think that a solution to not break stuff is just implement those kind of caching in this way we can improve the performance for all of the WP instances. Later we can evaluate the options we have discussed but atleast in this way we are doing something.

**Comment by aristath** [comment]
@dd32 
> I should also add - IMHO using `.php` files shouldn't be done as a "caching layer" but rather as a file format delivered by WordPress.org alongside the `.mo`, `.po` files and ultimately replacing them.

What if it's both? 🤔 
We could introduce a mechanism to use `.php` files for translations, and any plugins/themes that provide `.mo` files and not `.php` will be auto-converted and cached by Core.
Using this approach, we'd have a middle-ground: We'd be able to take advantage of the PHP format, while at the same time, we wouldn't be forcing plugins/themes to change immediately. If we detect a plugin/theme using the MO format without PHP, we can log a deprecation notice for developers so they can slowly start migrating to the new standard.

**Comment by bor0** [comment]
Some great ideas here - thanks for your thoughts!

To add to the ideas bucket :) There could be a "compilation" step so that translations are "compiled" from the original core PHP file into their own PHP file - basically, a simple substitution would do. So, instead of having `wp-includes/widgets.php`, it would be `wp-includes/de/widgets.php` where all `__( X )` calls are replaced with the corresponding string for `de` and similarly for `fr` etc.

This removes most performance bottlenecks (memory/CPU consumption or loading large files) and leaves it up to the filesystem/OS level to require the needed file.

As with any proposed idea, there are a couple of trade-offs/tricky bits here:

- This change has a high impact - we need to figure out tooling for translators, as well as when and how to recompile
  - Maybe we can use something like `stat` to check when the translated PHP file was created and whether the non-translated one has any modifications since
- Some instances will have a large number of files - `n*x` files where `n` is the number of available languages and `x` is the number of core PHP files. This will likely affect things like OPcache.
- As mentioned in another comment, not all hosting providers allow file system modification (if we were to compile these on the fly)
- Figuring out how to load `wp-includes/de/widgets.php` instead of `wp-includes/widgets.php` is another tricky bit - we would need to introduce an abstraction on top of `require*`/`include*`.

**Comment by swissspidy** [comment]
> There could be a "compilation" step so that translations are "compiled" from the original core PHP file into their own PHP file

Interesting idea, but I don't think it will be feasible. It would need to happen at build time on WordPress.org as you can't expect a single site to do this kind of modifications. Then in addition to the problem with a large number of files you also won't be able to do things like locale switching anymore because you can't redeclare functions/classes/constants.
It also doesn't solve the issue for plugins and themes.

-----

FWIW I am currently doing an in-depth analysis on this topic here, planning on sharing results soon!

**Comment by bor0** [comment]
> It would need to happen at build time on WordPress.org

My idea was for this build step to be a part of WordPress core itself, and do the "compilation" dynamically

**Comment by swissspidy** [comment]
ICYMI, I just published a thorough i18n performance analysis yesterday: https://make.wordpress.org/core/2023/07/24/i18n-performance-analysis/

It builds upon the recent [server performance analysis](https://make.wordpress.org/core/2023/05/25/wordpress-6-2-server-performance-analysis-summary/) to compare various solutions that have been discussed in the past, including the conversation here.

Looking forward to any comments on the post!

**Comment by swissspidy** [comment]
It has been two weeks since the blog post was published and the deadline for providing feedback just passed.

Overall the responses were constructive and positive. Nothing was brought up that hasn’t already been accounted for. I’d like to especially call out [this comment](https://make.wordpress.org/core/2023/07/24/i18n-performance-analysis/#comment-45287) by @Chouby, which I thought summarizes the situation quite well as we’re essentially dealing with a space-time tradeoff when trying to improve i18n performance in WordPress.

There was an interesting question about whether locales like Japanese have different memory consumption? [A quick test didn’t show any noticeable differences](https://github.com/swissspidy/wp-i18n-benchmarks/pull/13), and any improvements were consistent between the different locales, so it’s not like some locales would benefit more than others.

The blog post ended with the conclusion that a solution like [Ginger MO](https://github.com/swissspidy/ginger-mo) has the best horizontal and vertical impact and is thus worth looking into further. This sentiment was shared in the comments as well.

A point was raised about OPCache availability and configuration on shared hosting providers potentially not being well suited for PHP translation files. This is something we need to do more real life testing on.

Therefore, I’d suggest continuing development of Ginger MO and making it available as a feature plugin for people to test. This way, we can further refine the approach and get to that real life testing on common hosting providers. Hopefully we do not uncover any major blockers so that we could consider a merge proposal for WordPress 6.5 the earliest.

Let’s figure out the best approach for that together!

Since Performance Lab modules will be [split up into standalone plugins](https://github.com/WordPress/performance/issues/656) anyway, I personally think it makes sense for Ginger MO to remain a standalone plugin and repository. This way we can keep all the Git history in one place and do not have to migrate over the whole extensive codebase before having to migrate it again into core right afterwards. The Performance Lab plugin can still advertise Ginger MO as a module.

**Comment by swissspidy** [comment]
Here's a long overdue update!

[Performant Translations](https://wordpress.org/plugins/performant-translations/) (formerly known as Ginger MO) is now available as a standalone feature plugin in the plugin directory for wider testing. Once #656 is done it will be featured in the Performance Lab plugin accordingly.

I encourage everyone to check out the plugin and see for themselves how well it works.

Bugs can be filed at https://github.com/swissspidy/performant-translations

**Comment by Mte90** [comment]
I am following the development of that plugin, has the best optimizations and hints also from this ticket so I invite everyone involved to take a look.

**Comment by swissspidy** [comment]
With 6.5 I'd say this is complete now :-)
